New model support RTDETR (#29077)
huggingface/transformers: docs/source/en/model_doc/rt_detr.md

The model version was contributed by [rafaelpadilla](https://huggingface.co/rafaelpadilla) and [sanbumchoi](https://github.com/SangbumChoi). The original code can be found [here](https://github.com/lyuwenyu/RT-DETR/).

> 

Author: ```
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

image_processor = RTDetrImageProcessor()
inputs = image_processor(images=image, size={"height": 640, "width": 640}, return_tensors="pt")
label = False
if label:
    n_targets = 8
    batch_size = 2
    torch_device = 'cuda'
    labels = []
    for i in range(batch_size):
        target = {}
        target["class_labels"] = torch.ones(
            size=(n_targets,), device=torch_device, dtype=torch.long
        )
        target["boxes"] = torch.rand(
            n_targets, 4, device=torch_device, dtype=torch.float
        )
        labels.append(target)
    inputs['labels'] = labels
inputs["pixel_values"] = inputs["pixel_values"].cuda()
```
Something like this should be enought? (This is the part of the code that I shared in the slack)

> 

Reviewer: I would follow something like in our tests, which you can take a single COCO annotation and image, pass that to the image processor and then pass the processed inputs to the model

> 

Author: I have added like this

image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
with open("./tests/fixtures/tests_samples/COCO/coco_annotations.txt", "r") as f:
    target = json.loads(f.read())

target = {"image_id": 39769, "annotations": target}

image_processing = RTDetrImageProcessor.from_pretrained("sbchoi/rtdetr_r50vd")
encoding = image_processing(images=image, annotations=target, return_tensors="pt")
model = RTDetrForObjectDetection.from_pretrained("sbchoi/rtdetr_r50vd")
outputs = model(**encoding)

Is it okay?

> 

Reviewer: Perfect 👌

> 

Author: also I think this is github bug?
https://github.com/huggingface/transformers/blob/df99241dbb9b020c60151db0b3775a5da09c62cb/docs/source/en/model_doc/rt_detr.md

## Usage tips

RT-DETR is an object detection model that stands for "Real-Time DEtection Transformer." This model is designed to perform object detection tasks with a focus on achieving real-time performance while maintaining high accuracy. Leveraging the transformer architecture, which has gained significant popularity in various fields of deep learning, RT-DETR processes images to identify and locate multiple objects within them.

Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet-D variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a hybrid encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads to refine object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.

```

> image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")

Reviewer 2: Would be great to load the cats image like so:
```python
import requests
from PIL import Image

url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)
```

> image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")

Author: done!

<!--Copyright 2024 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

# RT-DETR

## Overview

> The RT-DETR model was proposed in [DETRs Beat YOLOs on Real-time Object Detection](https://arxiv.org/abs/2304.08069) by Wenyu Lv, Yian Zhao, Shangliang Xu, Jinman Wei, Guanzhong Wang, Cheng Cui, Yuning Du, Qingqing Dang, Yi Liu.

Reviewer 2: (nit) usually we add the abstract from the paper here, not in the "usage tips" section

The model version was contributed by [rafaelpadilla](https://huggingface.co/rafaelpadilla) and [sangbumchoi](https://github.com/SangbumChoi). The original code can be found [here](https://github.com/lyuwenyu/RT-DETR/).

## Usage tips

> Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet-D variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a hybrid encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads to refine object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.

Reviewer 2: ```suggestion
Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet-D variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a hybrid encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads is used to refine the object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.
```

```
from transformers import RTDetrForObjectDetection, RTDetrImageProcessor
from PIL import Image
import json
import torch

image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
with open("./tests/fixtures/tests_samples/COCO/coco_annotations.txt", "r") as f:
    target = json.loads(f.read())

target = {"image_id": 39769, "annotations": target}

> image_processing = RTDetrImageProcessor.from_pretrained("sbchoi/rtdetr_r50vd")
> encoding = image_processing(images=image, annotations=target, return_tensors="pt")
> model = RTDetrForObjectDetection.from_pretrained("sbchoi/rtdetr_r50vd")
> outputs = model(**encoding)
> results = image_processing.post_process_object_detection(outputs, target_sizes=torch.tensor([[480, 640]]), threshold=0.3)

Reviewer 2: ```suggestion
image_processing = RTDetrImageProcessor.from_pretrained("sbchoi/rtdetr_r50vd")
encoding = image_processing(images=image, annotations=target, return_tensors="pt")
model = RTDetrForObjectDetection.from_pretrained("sbchoi/rtdetr_r50vd")

with torch.no_grad():
   outputs = model(**encoding)

results = image_processing.post_process_object_detection(outputs, target_sizes=torch.tensor([[480, 640]]), threshold=0.3)
```

## Usage tips

Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet-D variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a hybrid encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads is used to refine the object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.

```py
from transformers import RTDetrForObjectDetection, RTDetrImageProcessor
from PIL import Image
import json
import torch

image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
with open("./tests/fixtures/tests_samples/COCO/coco_annotations.txt", "r") as f:
    target = json.loads(f.read())

target = {"image_id": 39769, "annotations": target}

> image_processing = RTDetrImageProcessor.from_pretrained("sbchoi/rtdetr_r50vd")
> encoding = image_processing(images=image, annotations=target, return_tensors="pt")
> model = RTDetrForObjectDetection.from_pretrained("sbchoi/rtdetr_r50vd")
> 
> with torch.no_grad():
>    outputs = model(**encoding)

Reviewer 2: ```suggestion
image_processor = RTDetrImageProcessor.from_pretrained("sbchoi/rtdetr_r50vd")
model = RTDetrForObjectDetection.from_pretrained("sbchoi/rtdetr_r50vd")

inputs = image_processor(images=image, annotations=target, return_tensors="pt")

with torch.no_grad():
   outputs = model(**inputs)
```

> results = image_processing.post_process_object_detection(outputs, target_sizes=torch.tensor([[480, 640]]), threshold=0.3)

Reviewer 2: ```suggestion
results = image_processor.post_process_object_detection(outputs, target_sizes=torch.tensor([image.size[::-1]), threshold=0.3)
```

## Usage tips

Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet-D variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a hybrid encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads is used to refine the object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.

```py
from transformers import RTDetrForObjectDetection, RTDetrImageProcessor
from PIL import Image
import json
import torch
import requests

url = 'http://images.cocodataset.org/val2017/000000039769.jpg' 
image = Image.open(requests.get(url, stream=True).raw)
with open("./tests/fixtures/tests_samples/COCO/coco_annotations.txt", "r") as f:
    target = json.loads(f.read())

> target = {"image_id": 39769, "annotations": target}

Reviewer 2: This can be removed since users are not running this from the Github repository

Reviewer 2: Perhaps you can show how to calculate a loss with a dummy dictionary

Author: I have removed the unncessary part which is target!
